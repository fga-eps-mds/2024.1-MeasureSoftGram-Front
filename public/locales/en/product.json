{
  "title-create": "ADD PRODUCT",
  "title-edit": "EDIT PRODUCT",
  "description": "Products are software that belong to an organization and have a client. Due to their nature, the same product can have several associated repositories, allowing continuous implementation across multiple work fronts.",
  "label-input": "Organizations",
  "title": "Products",
  "name-input": "Name",
  "search": "Search product",
  "description-input": "Description",
  "back": "BACK",
  "save": "SAVE",
  "create": "CREATE",
  "not-found": "No product found.",
  "toast": {
    "name-exists": "A Product with this name already exists.",
    "success-edit": "Product updated successfully!",
    "error-edit": "Error updating the Product!",
    "sucess": "Product created successfully!",
    "error-create": "Error creating the Product!"
  }
}
