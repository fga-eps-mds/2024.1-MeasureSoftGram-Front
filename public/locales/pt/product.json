{
  "title-create": "ADICIONAR PRODUTO",
  "title-edit": "EDITAR PRODUTO",
  "description": "Os produtos são softwares que pertencem à alguma organização e que possuem algum cliente. Por conta de sua natureza, um mesmo produto pode possuir vários repositórios associados à ele, o que permite uma implementação continua em mais de uma frente de trabalho.",
  "label-input": "Organizações",
  "name-input": "Nome",
  "description-input": "Descrição",
  "back": "VOLTAR",
  "save": "SALVAR",
  "create": "CRIAR",
  "toast": {
    "name-exists": "Já existe um Produto com este nome.",
    "success-edit": "Produto atualizado com sucesso!",
    "error-edit": "Erro ao atualizar o Produto!",
    "sucess": "Produto criado com sucesso!",
    "error-create": "Erro ao criar o Produto!"
  }
}
