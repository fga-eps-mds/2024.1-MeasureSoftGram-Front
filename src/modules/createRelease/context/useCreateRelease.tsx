import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { productQuery } from '@services/product';
import { Changes, Goal, PreConfigEntitiesRelationship, Product } from '@customTypes/product';
import { addDays, format } from 'date-fns';
import { mutate } from 'swr';
import { useRouter } from 'next/router';
import { Characteristic, Measure, PreConfigRoot, Subcharacteristic } from '@customTypes/preConfig';
import { entityRelationshipTreeService } from '@services/entityRelationshipTree';
import { CREATE_RELEASE_STEP } from '../consts';
import mockedData from '../utils/mockedData.json';
import { mergeCharacteristicData } from '../utils/mergeCharacteristicData';
import ReleaseGoals from '../components/ReleaseGoals';

interface ReleaseGoal {
  id: number;
  release_name: string;
  start_at: string;
  end_at: string;
  changes: Changes[];
  allow_dynamic: boolean;
}

interface NewCreateGoalData {
  changes: Changes[];
  allow_dynamic: boolean;
}

export interface NewCreateReleaseData {
  goal: number;
  description?: string;
  release_name: string;
  start_at: string;
  end_at: string;
}

interface CreateReleaseProviderProps {
  children: ReactNode;
  productId: string;
  organizationId: string;
  currentProduct: Product;
}

interface ReleaseInfoForm {
  name: string;
  description: string;
  startDate: string;
  endDate: string;
  characteristics: string[];
  changes: Changes[];
}

interface ConfigPageData {
  characteristicCheckbox: string[];
  setCharacteristicCheckbox: (characteristicCheckbox: string[]) => void;
  subcharacterCheckbox: string[];
  setSubcharacterCheckbox: (subcharacterCheckbox: string[]) => void;
  measureCheckbox: string[];
  setMeasureCheckbox: (measureCheckbox: string[]) => void;
  characteristicData: Characteristic[];
  setCharacteristicValuesValid: (value: boolean) => void;
}

interface CreateReleaseContextData {
  releaseInfoForm: ReleaseInfoForm;
  alertMessage: string;
  productId: string;
  organizationId: string;
  currentProduct: Product;
  isFirstRelease: boolean;
  configPageData: ConfigPageData;
  handleChangeForm: (field: string, value: string | Changes[]) => void;
  finishReleasePlanning: () => void;
  goToNextStep: (activeStep: number) => boolean;
  closeAlert: () => void;
  setCurrentConfig: (data: Characteristic[]) => void;
  changeThreshold: boolean;
  toggleChangeThreshold: () => void;
  getNextStep: (activeStep: number, configPage: number) => number;
  getPreviousStep: (activeStep: number, configPage: number) => number;
  allowChangeConfig: boolean;
  toggleAllowChangeConfig: () => void;
  setUseLastConfig: (value: boolean) => void;
  resetStates: () => void;
  lastGoal?: Goal;
  setAllowDynamicBalance: (value: boolean) => void;
}

export const CreateReleaseContext = createContext({} as CreateReleaseContextData);

const defaultEndDate = format(addDays(new Date(), 7), 'yyyy-MM-dd');
const defaultStartDate = format(new Date(), 'yyyy-MM-dd');

export function CreateReleaseProvider({
  children,
  productId,
  organizationId,
  currentProduct
}: CreateReleaseProviderProps) {
  const [alertMessage, setAlertMessage] = useState('');
  const [releaseInfoForm, setReleaseInfoForm] = useState<ReleaseInfoForm>({
    endDate: defaultEndDate,
    startDate: defaultStartDate,
    changes: [],
    characteristics: [],
    name: '',
    description: ''
  });
  const [characteristicData, setCharacteristicData] = useState<Characteristic[]>([]);
  const [characteristicCheckbox, setCharacteristicCheckbox] = useState<string[]>([]);
  const [subcharacterCheckbox, setSubcharacterCheckbox] = useState<string[]>([]);
  const [measureCheckbox, setMeasureCheckbox] = useState<string[]>([]);
  const [changeThreshold, setChangeThreshold] = useState<boolean>(false);
  const [allowChangeConfig, setAllowChangeConfig] = useState<boolean>(false);
  const [useLastConfig, setUseLastConfig] = useState<boolean>(false);
  const [isFirstRelease, setIsFirstRelease] = useState<boolean>(false);
  const [characteristicValuesValid, setCharacteristicValuesValid] = useState<boolean>(false)
  const [entityRelationshipTree, setEntityRelationshipTree] = useState<PreConfigEntitiesRelationship[]>([]);
  const [lastGoal, setLastGoal] = useState<Goal | undefined>(undefined);
  const [allowDynamicBalance, setAllowDynamicBalance] = useState<boolean>(false)

  const router = useRouter();

  async function checkForFirstRelease() {
    try {
      const result = await productQuery.getCurrentGoal(organizationId, productId);
      if (!result.data) setIsFirstRelease(true);
      setLastGoal(result.data as unknown as Goal);
    } catch (error) {
      console.error(error);
    }
  }

  async function createProductReleaseGoal() {
    if (useLastConfig) {
      const dataRelease: NewCreateReleaseData = {
        goal: lastGoal!.id,
        release_name: releaseInfoForm.name,
        start_at: releaseInfoForm.startDate,
        end_at: releaseInfoForm.endDate,
        description: releaseInfoForm.description || '',
      };

      const response = await productQuery.createProductRelease(organizationId, productId, dataRelease);
      await mutate(
        JSON.stringify({ url: `organizations/${organizationId}/products/${productId}/create/release/`, method: 'get' })
      );
      const releaseId = response?.data?.id;
      if (releaseId) {
        await router.push(
          `/products/${organizationId}-${productId}-${currentProduct.name.toLowerCase()}/releases/${releaseId}`
        );
        setAlertMessage('successOnCreation');
      }
      return;
    }

    const dataGoal: NewCreateGoalData = {
      changes: releaseInfoForm.changes,
      allow_dynamic: allowDynamicBalance
    };

    try {
      const responseGoal = await productQuery.createProductGoal(organizationId, productId, dataGoal as unknown as ReleaseGoal);

      const dataRelease: NewCreateReleaseData = {
        goal: responseGoal.data.id,
        release_name: releaseInfoForm.name,
        start_at: releaseInfoForm.startDate,
        end_at: releaseInfoForm.endDate,
        description: releaseInfoForm.description || '',
      };

      const responseRelease = await productQuery.createProductRelease(organizationId, productId, dataRelease);
      await mutate(
        JSON.stringify({ url: `organizations/${organizationId}/products/${productId}/create/release/`, method: 'get' })
      );

      const releaseId = responseRelease?.data?.id;

      if (releaseId) {
        await router.push(
          `/products/${organizationId}-${productId}-${currentProduct.name.toLowerCase()}/releases/${releaseId}`
        );
        setAlertMessage('successOnCreation');
      }
    } catch (error) {
      console.error(error);
    }
  }

  function handleChangeForm(field: string, value: string | string[] | Changes[]) {
    setReleaseInfoForm((form) => ({
      ...form,
      [field]: value
    }));
  }

  function goToNextStep(currentStep: number) {
    if (currentStep === CREATE_RELEASE_STEP.ReleaseInfoStep) {
      const { name, startDate, endDate } = releaseInfoForm;
      if (!name) {
        setAlertMessage('fillName');
        return false
      }
      if (Date.parse(startDate) > Date.parse(endDate)) {
        setAlertMessage('invalidDate');
        return false
      }
    }
    if (currentStep === CREATE_RELEASE_STEP.CharacteristicStep && !characteristicValuesValid) {
      setAlertMessage('characteristicValuesInvalid')
      return false;
    }
    return true;
  }

  function closeAlert() {
    setAlertMessage('');
  }

  function setCurrentConfig(configData: Characteristic[]) {
    const newConfigData = mergeCharacteristicData(configData, entityRelationshipTree);
    if (configData.length !== characteristicData.length && configData.length) {
      setCharacteristicCheckbox(configData.filter((c => c.weight > 0)).map((c: Characteristic) => c.key))
      const subCharacteristics = configData.flatMap((c: Characteristic) => c.subcharacteristics)
      setSubcharacterCheckbox(subCharacteristics.filter((sc => sc.weight > 0)).map((sc: Subcharacteristic) => sc.key))
      const measures = subCharacteristics.flatMap((sc: Subcharacteristic) => sc.measures)
      setMeasureCheckbox(measures.filter((m => m.weight > 0)).map((m: Measure) => m.key))
    }
    setReleaseInfoForm((form) => ({
      ...form,
      characteristics: configData.filter((c: Characteristic) => c.weight > 0).map((c: Characteristic) => c.key)
    }))
    setCharacteristicData(newConfigData);
  }

  async function loadCurrentConfig() {
    try {
      if (currentProduct) {
        const result = (await productQuery.getProductCurrentPreConfig(
          organizationId,
          productId
        )).data as unknown as PreConfigRoot;
        setCurrentConfig(result?.data?.characteristics ?? mockedData.data.characteristics);
      }
    } catch (error) {
      console.error(error);
    }
  }

  function toggleChangeThreshold() {
    setChangeThreshold(!changeThreshold);
  }

  function toggleAllowChangeConfig() {
    setAllowChangeConfig(!allowChangeConfig);
  }

  function getNextStep(activeStep: number, configPage: number) {
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep && configPage === 2) {
      return changeThreshold ? CREATE_RELEASE_STEP.ThresholdStep : CREATE_RELEASE_STEP.ReleaseGoalStep;
    }
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep) {
      return CREATE_RELEASE_STEP.CharacteristicStep;
    }
    return activeStep + 1;
  }

  function getPreviousStep(activeStep: number, configPage: number) {
    if (activeStep === CREATE_RELEASE_STEP.ReleaseGoalStep) {
      if (useLastConfig) {
        setUseLastConfig(false);
        return CREATE_RELEASE_STEP.ReleaseSelectorStep;
      }
      return changeThreshold ? CREATE_RELEASE_STEP.ThresholdStep : CREATE_RELEASE_STEP.CharacteristicStep;
    }
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep && configPage !== 0) {
      return CREATE_RELEASE_STEP.CharacteristicStep;
    }
    return activeStep - 1;
  }

  async function sendConfigJson() {
    const responseCharacterFiltered = characteristicData?.filter((charcterValue) =>
      characteristicCheckbox.includes(charcterValue.key)
    );
    const responseSubcharacterFiltered = responseCharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.filter((subcharcterValue) =>
        subcharacterCheckbox.includes(subcharcterValue.key)
      )
    }));
    const finalData = responseSubcharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.map((subcharcterValue) => ({
        ...subcharcterValue,
        measures: subcharcterValue.measures.filter((measureValue) => measureCheckbox.includes(measureValue.key))
      }))
    })) as Characteristic[];

    await productQuery
      .postPreConfig(organizationId, productId, { name: currentProduct?.name ?? '', data: { characteristics: finalData } })
  };

  function finishReleasePlanning() {
    sendConfigJson().catch(() => setAlertMessage('errorOnCreation'));
    createProductReleaseGoal().catch(() => setAlertMessage('errorOnCreation'));
  }
  async function loadEntityRelationshipTree() {
    try {
      const entityRelationshipTreeResult = (await entityRelationshipTreeService.getEntityRelationshipTree())
        .data as unknown as PreConfigEntitiesRelationship[];
      setEntityRelationshipTree(entityRelationshipTreeResult);
    } catch (error) {
      console.error(error);
    }
  }

  function resetStates() {
    setReleaseInfoForm({
      endDate: defaultEndDate,
      startDate: defaultStartDate,
      changes: [],
      characteristics: [],
      name: '',
      description: ''
    } as ReleaseInfoForm);
    setAllowChangeConfig(false);
    setChangeThreshold(false);
    loadEntityRelationshipTree();
    loadCurrentConfig();
    checkForFirstRelease();
  }

  useEffect(() => {
    resetStates()
  }, [productId]);

  // eslint-disable-next-line react/jsx-no-constructed-context-values
  const value = {
    releaseInfoForm,
    alertMessage,
    handleChangeForm,
    finishReleasePlanning,
    goToNextStep,
    closeAlert,
    productId,
    organizationId,
    currentProduct,
    isFirstRelease,
    setCurrentConfig,
    configPageData: {
      characteristicCheckbox,
      setCharacteristicCheckbox,
      subcharacterCheckbox,
      setSubcharacterCheckbox,
      characteristicData,
      setMeasureCheckbox,
      measureCheckbox,
      setCharacteristicValuesValid
    },
    changeThreshold,
    toggleChangeThreshold,
    getNextStep,
    getPreviousStep,
    allowChangeConfig,
    toggleAllowChangeConfig,
    setUseLastConfig,
    resetStates,
    lastGoal,
    setAllowDynamicBalance
  };

  return <CreateReleaseContext.Provider value={value}>{children}</CreateReleaseContext.Provider>;
}

export function useCreateReleaseContext(): CreateReleaseContextData {
  return useContext(CreateReleaseContext);
}
