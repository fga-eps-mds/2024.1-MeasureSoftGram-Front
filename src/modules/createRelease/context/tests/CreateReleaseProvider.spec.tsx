import '@testing-library/jest-dom';

import React, { ReactNode, useEffect } from 'react';
import { act, fireEvent, render, renderHook, waitFor } from '@testing-library/react';

import { Characteristics, Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { CreateReleaseProvider, useCreateReleaseContext } from '../useCreateRelease';

jest.mock('@services/product', () => ({
  productQuery: {
    getProductCurrentPreConfig: () => ({ data: { "id": 25, "name": "3", "data": { "characteristics": [{ "key": "reliability", "weight": 50, "subcharacteristics": [{ "key": "testing_status", "weight": 100, "measures": [{ "key": "passed_tests", "weight": 33 }, { "key": "test_builds", "weight": 33 }, { "key": "test_coverage", "weight": 34 }] }] }, { "key": "maintainability", "weight": 50, "subcharacteristics": [{ "key": "modifiability", "weight": 100, "measures": [{ "key": "non_complex_file_density", "weight": 33 }, { "key": "commented_file_density", "weight": 33 }, { "key": "duplication_absense", "weight": 34 }] }] }] }, "created_at": "2022-09-15T18:56:57-03:00" } }),
    getPreConfigEntitiesRelationship: () => ({
      data: [{
        key: 'reliability',
        id: 1,
        name: 'Confiabilidade',
        description: '',
        subcharacteristics: []
      }]
    }),
    getReleaseGoalList: () => ({ data: [{ id: 1, release_name: 'r1', start_at: '', end_at: '', changes: [] }] }),
    getCurrentGoal: () => ({ data: { id: 1, release_name: 'r1', start_at: '', end_at: '', changes: [] } }),
    createProductReleaseGoal: () => ({ data: { id: 1 } })
  }
}));

interface TestingComponentProps {
  step?: number
  configPage?: number
  startDate?: string
  endDate?: string
}

interface Props {
  children: ReactNode;
}

const Provider = ({ children }: Props) => <CreateReleaseProvider organizationId='1' productId='3' currentProduct={product}>{children}</CreateReleaseProvider>;


const TestingComponent = ({ step = 0, configPage = 0, startDate = '30/08/2022', endDate = '30/08/2022' }: TestingComponentProps) => {
  const {
    handleChangeForm,
    closeAlert,
    goToNextStep,
    finishReleasePlanning,
    setCurrentConfig,
    getPreviousStep,
    getNextStep,
    setUseLastConfig,
    toggleChangeThreshold,
    configPageData,
  } = useCreateReleaseContext();

  useEffect(() => {
    handleChangeForm('endDate', endDate)
    handleChangeForm('startDate', startDate)
    handleChangeForm('name', 'aoba')
    setCurrentConfig([])
  }, [])

  return (
    <>
      <h1>aoba</h1>
      <button
        type='submit'
        data-testid="close-alert"
        onClick={closeAlert}
      >Fechar alerta</button>
      <button
        type='submit'
        data-testid="goal-step"
        onClick={() => goToNextStep(step)}
      >Denifir pesos</button>
      <button
        type='submit'
        onClick={finishReleasePlanning}
      >Cria metas</button>
      <button
        type='submit'
        data-testid="test-previous-step"
        onClick={() => getPreviousStep(step, configPage)}
      >PreviousStep</button>
      <button
        type='submit'
        data-testid="test-next-step"
        onClick={() => getNextStep(step, configPage)}
      >NextStep</button>
      <button
        type='submit'
        data-testid="test-use-last-config"
        onClick={() => setUseLastConfig(true)}
      >UseLastConfig</button>
      <button
        type='submit'
        data-testid="test-change-threshold"
        onClick={toggleChangeThreshold}
      >ToggleChangeThreshold</button>
      <button
        type='submit'
        data-testid="test-set-characteristic-values-valid"
        onClick={() => configPageData.setCharacteristicValuesValid(false)}
      >SetIsValuesInvalid</button>
      <button
        type='submit'
        data-testid="test-set-invalid-date"
        onClick={() => {
          handleChangeForm('startDate', '30/08/2022')
          handleChangeForm('endDate', '29/08/2022')
          handleChangeForm('name', 'aoba')
        }}
      >SetInvalidDate</button>
    </>
  );
};

const product: Product = {
  id: "1",
  name: "MeasureSoftGram",
  description: '',
  github_url: '',
  created_at: '',
  updated_at: '',
  gaugeRedLimit: 0,
  gaugeYellowLimit: 0
}

const NextStepTestId = "test-next-step";
const PreviousStepTestId = "test-previous-step";

describe('<CreateReleaseProvider />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <CreateReleaseProvider
          currentProduct={product}
          organizationId='1'
          productId='3'
        >
          <TestingComponent />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeInTheDocument()
      })

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Comportamento', () => {
    it('Deve resetar estados', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId('close-alert'));
      fireEvent.click(tree.getByTestId('goal-step'));
    });

    describe('GetPreviousStep e GetNextStep', () => {
      it('Deve chamar getPreviousStep e GetNextStep corretamente para step 0', () => {
        const tree = render(
          <CreateReleaseProvider
            organizationId='1'
            productId='3'
            currentProduct={product}
          >
            <TestingComponent />
          </CreateReleaseProvider>
        )

        waitFor(() => {
          expect(tree.getAllByText('aoba')).toBeDefined()
        })

        fireEvent.click(tree.getByTestId(PreviousStepTestId));
        fireEvent.click(tree.getByTestId(NextStepTestId));
      });
    })
    it('Deve chamar getPreviousStep e GetNextStep corretamente para step 2 e configPage 0', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={2} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId(PreviousStepTestId));
      fireEvent.click(tree.getByTestId(NextStepTestId));
    })
    it('Deve chamar getPreviousStep e GetNextStep corretamente para step 2 e configPage 1', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={2} configPage={1} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId(PreviousStepTestId));
      fireEvent.click(tree.getByTestId(NextStepTestId));
    });
    it('Deve chamar getPreviousStep e GetNextStep corretamente para step 2 e configPage 2', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={2} configPage={2} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId(PreviousStepTestId));
      fireEvent.click(tree.getByTestId(NextStepTestId));
    });
    it('Deve chamar o getPreviousStep corretamente para step 4 usando ultima config', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={4} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId('test-use-last-config'));
      fireEvent.click(tree.getByTestId(PreviousStepTestId));
    })
    it('Deve chamar o getPreviousStep corretamente para step 4 sem usar ultima config e mudando threshold', async () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={4} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId('test-change-threshold'));
      fireEvent.click(tree.getByTestId(PreviousStepTestId));
    });
    it('Deve chamar o getPreviousStep corretamente para step 4 sem usar ultima config e sem mudar threshold', async () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={4} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId(PreviousStepTestId));
    });

    it('Deve chamar o getNextStep corretamente para step 2 e configPage 2 usando threshold', async () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={2} configPage={2} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

      fireEvent.click(tree.getByTestId('test-change-threshold'));
      fireEvent.click(tree.getByTestId(NextStepTestId));
    })

    it('Deve chamar o goToNextStep para o step de caracter√≠stica com valores inv√°lidos', async () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent step={2} />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })
      fireEvent.click(tree.getByTestId('test-change-threshold'));
      fireEvent.click(tree.getByTestId('goal-step'));
    })
    const dataa: Characteristics = {
      maintainability: undefined,
      reliability: undefined,
    }

    it('Deve chamar o finishReleasePlanning passando o mesmo releaseGoal', async () => {
      jest.spyOn(productQuery, 'getCurrentGoal').mockResolvedValueOnce({
        data: {
          id: 1,
          release_name: 'aoba',
          start_at: '2022-08-30',
          end_at: '2022-08-30',
          data: dataa as any
        } as any
      } as any)
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

    })

    it('Deve chamar o finishReleasePlanning passando um novo releaseGoal', async () => {
      jest.spyOn(productQuery, 'createProductReleaseGoal').mockResolvedValueOnce({
        data: {
          id: 1
        }
      } as any)
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
          currentProduct={product}
        >
          <TestingComponent />
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('aoba')).toBeDefined()
      })

    })

    it('Deve chamar o goToNextStep com datas inv√°lidas', async () => {
      const { result } = renderHook(() => useCreateReleaseContext(), { wrapper: Provider });

      act(() => {
        result.current.handleChangeForm('startDate', '2022-08-30')
        result.current.handleChangeForm('endDate', '2022-08-29')
        result.current.handleChangeForm('name', 'aoba')
      })

      act(() => {
        result.current.goToNextStep(0)
      })

      expect(result.current.alertMessage).toBe('invalidDate')
    });

    it('Deve setar isFirstRelease como true quando n√£o tiver release', async () => {
      jest.spyOn(productQuery, 'getCurrentGoal').mockResolvedValueOnce({
        data: null
      } as any)

      const { result } = renderHook(() => useCreateReleaseContext(), { wrapper: Provider });

      await waitFor(() => {
        expect(result.current.isFirstRelease).toBeTruthy()
      })
    });


    it('Deve setar lastgoal corretamente quando tiver release', async () => {
      jest.spyOn(productQuery, 'getCurrentGoal').mockResolvedValueOnce({
        data: {
          id: 1,
          release_name: 'aoba',
          start_at: '2022-08-30',
          end_at: '2022-08-30'
        }
      } as any)

      const { result } = renderHook(() => useCreateReleaseContext(), { wrapper: Provider });

      await waitFor(() => {
        expect(result.current.lastGoal).toEqual({
          id: 1,
          release_name: 'aoba',
          start_at: '2022-08-30',
          end_at: '2022-08-30'
        })
      })
    });
  });
});
