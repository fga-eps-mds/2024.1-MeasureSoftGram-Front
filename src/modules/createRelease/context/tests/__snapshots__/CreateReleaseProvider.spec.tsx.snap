// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateReleaseProvider /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        aoba
      </h1>
      <button
        data-testid="close-alert"
        type="submit"
      >
        Fechar alerta
      </button>
      <button
        data-testid="goal-step"
        type="submit"
      >
        Denifir pesos
      </button>
      <button
        type="submit"
      >
        Cria metas
      </button>
      <button
        data-testid="test-previous-step"
        type="submit"
      >
        PreviousStep
      </button>
      <button
        data-testid="test-next-step"
        type="submit"
      >
        NextStep
      </button>
      <button
        data-testid="test-use-last-config"
        type="submit"
      >
        UseLastConfig
      </button>
      <button
        data-testid="test-change-threshold"
        type="submit"
      >
        ToggleChangeThreshold
      </button>
      <button
        data-testid="test-set-characteristic-values-valid"
        type="submit"
      >
        SetIsValuesInvalid
      </button>
      <button
        data-testid="test-set-invalid-date"
        type="submit"
      >
        SetInvalidDate
      </button>
    </div>
  </body>,
  "container": <div>
    <h1>
      aoba
    </h1>
    <button
      data-testid="close-alert"
      type="submit"
    >
      Fechar alerta
    </button>
    <button
      data-testid="goal-step"
      type="submit"
    >
      Denifir pesos
    </button>
    <button
      type="submit"
    >
      Cria metas
    </button>
    <button
      data-testid="test-previous-step"
      type="submit"
    >
      PreviousStep
    </button>
    <button
      data-testid="test-next-step"
      type="submit"
    >
      NextStep
    </button>
    <button
      data-testid="test-use-last-config"
      type="submit"
    >
      UseLastConfig
    </button>
    <button
      data-testid="test-change-threshold"
      type="submit"
    >
      ToggleChangeThreshold
    </button>
    <button
      data-testid="test-set-characteristic-values-valid"
      type="submit"
    >
      SetIsValuesInvalid
    </button>
    <button
      data-testid="test-set-invalid-date"
      type="submit"
    >
      SetInvalidDate
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
